src/main/java/frc/trigon/robot/poseestimation/apriltagcamera/AprilTagCamera.java

```diff
@@ -0,0 +1,206 @@
+     *
+     * @param distanceFromBestTag the average of the distance from the best visible tag
+     * @return the robot's pose
+     */
+    private Pose2d calculateBestRobotPose(double distanceFromBestTag) {
+        if (isWithinBestTagRangeForSolvePNP())
+            return cameraPoseToRobotPose(inputs.solvePNPPose);
+        return calculateAssumedRobotHeadingPose().toPose2d();
+    }
+
+    /**
+     * Calculates the robot's pose by assuming its heading, the apriltag's position, and the camera's position on the robot.
+     * This method of finding the robot's pose is more robust but relies on knowing the robot's heading beforehand.
+     *
+     * @return the robot's pose
+     */
+    private Pose3d calculateAssumedRobotHeadingPose() {
+        final Rotation2d robotHeadingAtResultTimestamp = PoseEstimator6328.getInstance().samplePose(inputs.latestResultTimestampSeconds).getRotation();
+        final Translation2d robotFieldRelativePositionTranslation = getRobotFieldRelativePosition(robotHeadingAtResultTimestamp);
+        return new Pose3d(new Pose2d(robotFieldRelativePositionTranslation, robotHeadingAtResultTimestamp));
+    }
+
+    private Translation2d getRobotFieldRelativePosition(Rotation2d robotHeading) {
+        final Pose3d tagPose = FieldConstants.TAG_ID_TO_POSE.get(inputs.visibleTagIDs[0]);
+        final Translation2d tagToCamera = getTagToCamera(tagPose);
+        final Translation2d robotToTag = tagToCamera.plus(robotCenterToCamera.getTranslation().toTranslation2d());
+        final Translation2d fieldRelativeTagPose = robotToTag.rotateBy(robotHeading);
+        return tagPose.getTranslation().toTranslation2d().plus(fieldRelativeTagPose);
+    }
+
+    private Translation2d getTagToCamera(Pose3d tagPose) {
+        final double cameraToTagDistanceMeters = -PhotonUtils.calculateDistanceToTargetMeters(
+                robotCenterToCamera.getZ(), tagPose.getZ(), robotCenterToCamera.getRotation().getY(), inputs.bestTargetRelativePitchRadians
+        );
+        final double cameraToTagXDistance = Math.sin(inputs.bestTargetRelativeYawRadians + robotCenterToCamera.getRotation().getZ()) * cameraToTagDistanceMeters;
+        final double cameraToTagYDistance = -Math.cos(inputs.bestTargetRelativeYawRadians + robotCenterToCamera.getRotation().getZ()) * cameraToTagDistanceMeters;
+        return new Translation2d(cameraToTagXDistance, cameraToTagYDistance);
+    }
+
+    private Pose2d cameraPoseToRobotPose(Pose3d cameraPose) {
+        if (cameraPose == null)
+            return null;
+        return cameraPose.transformBy(robotCenterToCamera.inverse()).toPose2d();
+    }
+
+    private boolean isNewTimestamp() {
+        if (lastUpdatedTimestamp == getLatestResultTimestampSeconds())
+            return false;
+
+        lastUpdatedTimestamp = getLatestResultTimestampSeconds();
+        return true;
+    }
+
+    private void logVisibleTags() {
```

Comment:
Move logging stuff to the bottom of the class, and this is considered excessive logging.